### PR TITLE
share and protect API Status struct

### PR DESCRIPTION
**Pull request by dajohi**
With appContext, protect with its own mutex.
With insightApiContext, no need for a mutex since it is never
modified after creation.

### CODE REVIEW

File: api/apiroutes.go [hunk 1]
```diff
@@ -1454,8 +1459,12 @@ func (c *appContext) addressIoCsv(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	c.statusMtx.RLock()
+	statusHeight := c.Status.Height
+	c.statusMtx.RUnlock()
+
 	filename := fmt.Sprintf("address-io-%s-%d-%s.csv", address,
-		c.Status.GetHeight(), strconv.FormatInt(time.Now().Unix(), 10))
+		statusHeight, strconv.FormatInt(time.Now().Unix(), 10))
```
Comment (chappjc) [inline api/apiroutes.go]: This is actually why I avoid adding named mutexes that apply to a specific field of the struct.  Having the Status type have it's own mutex makes ensures the mutex is only used for Status accesses.  Plus the `GetHeight()` method does the locking for you, and there's no temp variable.  All around it's cleaner IMO.

File: api/insight/apiroutes.go [hunk 2]
```diff
@@ -472,7 +472,7 @@ func (c *insightApiContext) getTransactions(w http.ResponseWriter, r *http.Reque
 		}
 		addresses := []string{address}
 		rawTxs, recentTxs, err :=
-			c.BlockData.ChainDB.InsightAddressTransactions(addresses, int64(c.Status.GetHeight()-2))
+			c.BlockData.ChainDB.InsightAddressTransactions(addresses, int64(c.Status.Height-2))
```
Comment (chappjc) [inline api/insight/apiroutes.go]: It's definitely a bug with `insightApiContext` never updates any of the status fields.  The idea is probably to have the `insightApiContext` have a pointer to the same Status in `appContext`.

File: api/apiroutes.go [hunk 3]
```diff
@@ -338,7 +339,11 @@ func (c *appContext) coinSupply(w http.ResponseWriter, r *http.Request) {
 
 func (c *appContext) currentHeight(w http.ResponseWriter, _ *http.Request) {
 	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
-	if _, err := io.WriteString(w, strconv.Itoa(int(c.Status.GetHeight()))); err != nil {
+	c.statusMtx.RLock()
+	statusHeight := c.Status.Height
+	c.statusMtx.RUnlock()
+
+	if _, err := io.WriteString(w, strconv.Itoa(int(statusHeight))); err != nil {
```
Comment (chappjc) [inline api/apiroutes.go]: Like the other example, this is far less clean.

File: api/apiroutes.go (hunk 1 above)
Comment (dajohi) [reply api/apiroutes.go]: The fields are exported so the API allows you to bypass the function calls that keep data integrity.  Perhaps a third solution is needed :)

Comment (chappjc) [reply api/apiroutes.go]: True, there is ample opportunity for messing with the data the way it is set up.  Unexporting the fields and making a setter method is probably the third method then.

**Comment by chappjc** [comment]
As per the [review discussion](https://github.com/decred/dcrdata/pull/1072#discussion_r261841262), the plan of action is now to unexport the status fields, and make thread safe accessors.  There was already a getter, leaving a needed setter to be implemented.  Finally, sharing the status struct between dcrdata api and insight api is needed.

**Comment by chappjc** [comment]
Closing in favor of a future PR addressing the related issue https://github.com/decred/dcrdata/issues/1094

**Pull request closed without merging**